runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/AbstractWindmillStream.java

```diff
@@ -278,24 +335,73 @@ public String backendWorkerToken() {
   }
 
   @Override
-  public void shutdown() {
+  public final void shutdown() {
+    // Don't lock here as isShutdown checks are used in the stream to free blocked
+    // threads or as exit conditions to loops.
     if (isShutdown.compareAndSet(false, true)) {
       requestObserver()
           .onError(new WindmillStreamShutdownException("Explicit call to shutdown stream."));
+      shutdownInternal();
+      shutdownTime.set(DateTime.now());
     }
   }
 
-  private void setLastError(String error) {
-    lastError.set(error);
-    lastErrorTime.set(DateTime.now());
+  private void recordRestartReason(String error) {
+    lastRestartReason.set(error);
+    lastRestartTime.set(DateTime.now());
   }
 
+  protected abstract void shutdownInternal();
+
   public static class WindmillStreamShutdownException extends RuntimeException {
     public WindmillStreamShutdownException(String message) {
       super(message);
     }
   }
 
+  /**
+   * Request observer that allows resetting its internal delegate using the given {@link
+   * #requestObserverSupplier}.
+   */
+  @ThreadSafe
+  private static class ResettableRequestObserver<RequestT> implements StreamObserver<RequestT> {
+
+    private final Supplier<StreamObserver<RequestT>> requestObserverSupplier;
+
+    @GuardedBy("this")
+    private @Nullable StreamObserver<RequestT> delegateRequestObserver;
+
+    private ResettableRequestObserver(Supplier<StreamObserver<RequestT>> requestObserverSupplier) {
```

Comment:
```suggestion
    private ResettableRequestObserver(Supplier<StreamObserver<RequestT>> requestObserverSupplier) {
```
```suggestion
    public ResettableRequestObserver(Supplier<StreamObserver<RequestT>> requestObserverSupplier) {
```

make it public since it is used outside the private class?